src/v/datalake/proto_to_arrow_struct.h

```diff
@@ -0,0 +1,135 @@
+#pragma once
+
+#include "datalake/proto_to_arrow_interface.h"
+#include "datalake/proto_to_arrow_scalar.h"
+
+#include <arrow/api.h>
+#include <arrow/type.h>
+#include <google/protobuf/message.h>
+
+#include <memory>
+#include <stdexcept>
+
+namespace datalake::detail {
+
+class proto_to_arrow_struct : public proto_to_arrow_interface {
+public:
+    explicit proto_to_arrow_struct(
+      const google::protobuf::Descriptor* message_descriptor) {
+        using namespace detail;
+        namespace pb = google::protobuf;
+
+        // Set up child arrays
+        for (int field_idx = 0; field_idx < message_descriptor->field_count();
+             field_idx++) {
+            auto desc = message_descriptor->field(field_idx);
+            if (desc->cpp_type() == pb::FieldDescriptor::CPPTYPE_INT32) {
+                _child_arrays.push_back(
+                  std::make_unique<proto_to_arrow_scalar<arrow::Int32Type>>());
+            } else if (desc->cpp_type() == pb::FieldDescriptor::CPPTYPE_INT64) {
+                _child_arrays.push_back(
+                  std::make_unique<proto_to_arrow_scalar<arrow::Int64Type>>());
+            } else if (desc->cpp_type() == pb::FieldDescriptor::CPPTYPE_BOOL) {
+                _child_arrays.push_back(
+                  std::make_unique<
+                    proto_to_arrow_scalar<arrow::BooleanType>>());
+            } else if (desc->cpp_type() == pb::FieldDescriptor::CPPTYPE_FLOAT) {
+                _child_arrays.push_back(
+                  std::make_unique<proto_to_arrow_scalar<arrow::FloatType>>());
+            } else if (
+              desc->cpp_type() == pb::FieldDescriptor::CPPTYPE_DOUBLE) {
+                _child_arrays.push_back(
+                  std::make_unique<proto_to_arrow_scalar<arrow::DoubleType>>());
+            } else if (
+              desc->cpp_type() == pb::FieldDescriptor::CPPTYPE_STRING) {
+                _child_arrays.push_back(
+                  std::make_unique<proto_to_arrow_scalar<arrow::StringType>>());
+            } else if (
+              desc->cpp_type() == pb::FieldDescriptor::CPPTYPE_MESSAGE) {
+                auto field_message_descriptor = desc->message_type();
+                if (field_message_descriptor == nullptr) {
+                    throw std::runtime_error(
```

Comment:
wondering a std::error-code (integer) like the RPC layer makes more sense here, since this is likely to be called in hot paths and exceptions/runtimes are expensive with global lock acquisitions.